packages/valory/skills/abstract_round_abci/tests/test_tools/test_rounds.py

```diff
@@ -0,0 +1,689 @@
+            exit_event,
+            finished,
+        )
+        for _ in range(4):
+            next(gen)
+        last_iteration(gen)
+
+
+class DummyCollectSameUntilThresholdRoundWithEndBlock(
+    DummyCollectSameUntilThresholdRound
+):
+    """A `DummyCollectSameUntilThresholdRound` with `end_block` overriden."""
+
+    def __init__(self, dummy_exit_event: DummyEvent, *args: Any, **kwargs: Any):
+        """Initialize the dummy class."""
+        super().__init__(*args, **kwargs)
+        self.dummy_exit_event = dummy_exit_event
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Enum]]:
+        """A dummy `end_block` override."""
+        if self.threshold_reached:
+            return (
+                cast(
+                    DummySynchronizedData,
+                    self.synchronized_data.update(
+                        most_voted_keeper_address=self.most_voted_payload
+                    ),
+                ),
+                self.dummy_exit_event,
+            )
+        if not self.is_majority_possible(
+            self.collection, self.synchronized_data.nb_participants
+        ):
+            return self.synchronized_data, DummyEvent.NO_MAJORITY
+        return None
+
+
+class TestBaseCollectSameUntilThresholdRoundTest(BaseTestBase):
+    """Test `BaseCollectSameUntilThresholdRoundTest`."""
+
+    base_round_test: BaseCollectSameUntilThresholdRoundTest
+    base_round_test_cls: Type[
+        BaseCollectSameUntilThresholdRoundTest
+    ] = BaseCollectSameUntilThresholdRoundTest
+
+    @given(
+        st.sampled_from(DummyEvent),
+        st.binary(),
+    )
+    def test_test_round(
+        self, exit_event: DummyEvent, most_voted_payload: bytes
+    ) -> None:
+        """Test `_test_round`."""
+        self.base_round_test.consensus_params._max_participants = (  # pylint: disable=protected-access
+            MAX_PARTICIPANTS
+        )
+        test_round = DummyCollectSameUntilThresholdRoundWithEndBlock(
+            exit_event,
+            self.base_round_test.synchronized_data,
+            self.base_round_test.consensus_params,
+        )
+        round_payloads = {
+            f"test{i}": DummyTxPayload(f"agent_{i}", most_voted_payload)
+            for i in range(MAX_PARTICIPANTS)
+        }
+        synchronized_data_attr_checks = [
+            lambda _synchronized_data: _synchronized_data.most_voted_keeper_address
+        ]
+
+        # Create `_test_round` generator and trigger for the first time.
+        gen = self.base_round_test._test_round(  # pylint: disable=protected-access
+            test_round,
+            round_payloads,
+            lambda synchronized_data, _: synchronized_data.update(
+                most_voted_keeper_address=most_voted_payload
+            ),
+            synchronized_data_attr_checks,
+            most_voted_payload,
+            exit_event,
+        )
+        for _ in range(4):
+            next(gen)
+        last_iteration(gen)
```

Comment:
this is also repeated 5 times. Put into a function and place it on the base class

Reply:
Refactored in 2502ecd562d7bb198078e9edb96444bdb505266e.